hedera-node/hedera-app/src/main/java/com/hedera/node/app/info/StateNetworkInfo.java

```diff
@@ -56,7 +56,14 @@ public class StateNetworkInfo implements NetworkInfo {
      */
     private final Roster activeRoster;
 
-    private final Map<Long, NodeInfo> nodeInfos;
+    /**
+     * Non-final because we need {@code handleTransaction} to be able to swap in an
+     * updated map atomically, without giving a pre-handle thread a temporary view of
+     * an empty map. (Note that {@code handleTransaction}'s updates will only change
+     * <i>metadata</i> of nodes, but not the set of nodes itself; so pre-handle threads
+     * can use any version of the map to test for address book membership.)
+     */
+    private Map<Long, NodeInfo> nodeInfos;
```

Comment:
I suggest to make this field `volatile`. It probably does not matter because the pre-handle threads can use any version as you write, but it would make it more obvious that this field is accessed by multiple threads.